### PR TITLE
Improve the breadcrumb navigation for the exam exercise tutor dashboard

### PR DESCRIPTION
**Pull request by FrankeLukas**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I followed the [coding and design guidelines](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html).
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)

### Motivation and Context

Currently going to the exercise tutor dashboard brings you from the exam to the course, breaking breadcrumb navigation.

### Description

Moves the routes to have the dashboard as child route of the exam. Also fixes a crash when opening the Assessment Dashboard when there are no exercise groups.

### Steps for Testing

1. Navigate to an exam and visit the Assessment Dashboard
2. Open an Exercise Dashboard from there
3. Make sure no errors occur and you can still use the dashboard as usual
4. Open the Assessment Dashboard in an exam without exercise groups and check that there are no errors
5. Open Tutor and Instructor Assessment Dashboards of a course and make sure they work as on develop

### Screenshots

The breadcrumbs now fully display the exam route instead of a course route:
grafik

### CODE REVIEW

**Review by airelawaleria — COMMENTED** [review]
is it related to this pr that the routes underlined with red throw "Error: Cannot match any routes" in the console? Otherwise assessment and exercise dashboards worked fine
image

**Review by JohannesWeiss — COMMENTED** [review]
Got the same errors on the console as @airelawaleria.
image
Aside from that, no issues found.

**Comment by FrankeLukas** [comment]
Yes, that is expected right now (which isn't great), but will go away once #2873 is merged.

**Review by marvinchimeka — DISMISSED** [review]
Tested on ts3. Works as expected 👍

**Review by simonzilker — DISMISSED** [review]
Looks good on ts3

**Review by stefanwaldhauser — APPROVED** [review]
Works and code looks fine too

File: src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts [hunk 1]
```diff
@@ -144,7 +144,7 @@ export class ExerciseAssessmentDashboardComponent implements OnInit, AfterViewIn
     ngOnInit(): void {
         this.exerciseId = Number(this.route.snapshot.paramMap.get('exerciseId'));
         this.courseId = Number(this.route.snapshot.paramMap.get('courseId'));
-        this.isTestRun = this.route.snapshot.url[3]?.toString() === 'test-run-exercise-assessment-dashboard';
+        this.isTestRun = this.route.snapshot.url.length > 8 && this.route.snapshot.url[8]?.toString() === 'test-run-exercise-assessment-dashboard';
```
Comment (iozbeyli) [inline src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:147]: I don't know how but can this be done without just writing 8 there?

Comment (FrankeLukas) [reply src/main/webapp/app/exercises/shared/dashboards/tutor/exercise-assessment-dashboard.component.ts:147]: By using the router perhaps. I didn't do much here to reduce the risk of introducing regressions. Tbh, if we want to change this we should probably think of a way to determine if this is a test run without relying on some magic values in the url.

**Review by airelawaleria — APPROVED** [review]

**Comment by krusche** [comment]
I think a different route would make more sense (also in the context of courses)

## Course

### Assessment Dashboard
current: https://artemis.ase.in.tum.de/course-management/{courseId}/tutor-dashboard
better: https://artemis.ase.in.tum.de/course-management/{courseId}/assessment-dashboard

### Exercise Assessment Dashboard
current: https://artemis.ase.in.tum.de/course-management/{courseId}/exercises/{exerciseId}/tutor-dashboard
better: https://artemis.ase.in.tum.de/course-management/{courseId}/assessment-dashboard/{exerciseId} 

## Exam

### Assessment Dashboard
current: https://artemis.ase.in.tum.de/course-management/{courseId}/exams/{examId}/tutor-exam-dashboard
better: https://artemis.ase.in.tum.de/course-management/{courseId}/exams/{examId}/assessment-dashboard

### Exercise Assessment Dashboard
current: https://artemis.ase.in.tum.de/course-management/{courseId}/exercises/{exerciseId}/tutor-dashboard
proposed: https://artemis.ase.in.tum.de/course-management/{courseId}/exams/{examId}/exercise-groups/{exercise-group}/exercises/{exerciseId}/tutor-dashboard
better: https://artemis.ase.in.tum.de/course-management/{courseId}/exams/{examId}/assessment-dashboard/{exerciseId} 


Also notice the renaming in the URL from `tutor-dashboard` and `tutor-exam-dashboard` to `assessment-dashboard`

Can you please change it to the `better` version above?
This would be more intuitive in terms of navigation

**Review by krusche — DISMISSED** [review]
see my previous comment

File: src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard.component.html [hunk 2]
```diff
@@ -164,24 +164,15 @@ <h2 *ngIf="isExamMode">
 
                     <td>
                         <ng-container *ngIf="!exercise.teamMode; else teamsLink">
-                            <ng-container *ngIf="!isTestRun; else testRun">
+                            <ng-container>
                                 <a
-                                    [routerLink]="['/course-management', courseId, 'exercises', exercise.id, 'tutor-dashboard']"
+                                    [routerLink]="getTutorDashboardLinkForExercise(exercise)"
```
Comment (SimonEntholzer) [inline src/main/webapp/app/course/dashboards/assessment-dashboard/assessment-dashboard.component.html:169]: I guess this method should be renamed to `getTutorDashboardLinkForExercise(exercise)`

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
I actually think it's a really smart move to the exercise-assessment-dashboard to assessment-dashboard, and just appending the exercise id. Though, I'm not really convinced if the exercise-groups part in the url makes a lot of sense, but I guess that's debatable.
image
image

Other than that everything I tested worked great, and the code looks good aswell :D 
**edit:** just realized my requested change was already requested

File: src/main/webapp/app/exam/manage/exam-management.route.ts [hunk 3]
```diff
@@ -454,14 +455,34 @@ export const examManagementRoute: Routes = [
...
+        },
+        canActivate: [UserRouteAccessService],
+    },
+    {
+        path: ':examId/exercise-groups/:groupId/test-assessment-dashboard/:exerciseId',
+        component: ExerciseAssessmentDashboardComponent,
+        data: {
+            authorities: [Authority.ADMIN, Authority.INSTRUCTOR],
+            usePathForBreadcrumbs: true,
+            pageTitle: 'artemisApp.exerciseAssessmentDashboard.home.title',
```
Comment (derLalla) [inline src/main/webapp/app/exam/manage/exam-management.route.ts:482]: we might want to include `"Test"` to the title?

**Review by derLalla — CHANGES_REQUESTED** [review]
The code changes are really good. I have only left one small remark regarding a title

**Review by SimonEntholzer — APPROVED** [review]
Tested on TS1, works as expected. The new code changes are fine too.

**Review by tiffca — APPROVED** [review]
Tested successfully

**Review by EneaGore — COMMENTED** [review]
It generally looks good but is it intentional that the tutor and intrsuctor breadcumbs are different in the case of Course>Exam>Dashboard like in the picture. On the left tutor has Id of exam and to the right instructor has name of Exam.
tutorleft_instructorright

**Review by ivanchimeno — APPROVED** [review]
Code looks good, I approve :+1:

**Review by airelawaleria — APPROVED** [review]

**Pull request merged**